File: web/packages/teleport/src/services/integrations/integrations.test.ts
/**
 * Teleport
 * Copyright (C) 2023  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import cfg from 'teleport/config';
import { AwsResource } from 'teleport/Integrations/status/AwsOidc/StatCard';
import api from 'teleport/services/api';

import { integrationService } from './integrations';
import {
  IntegrationAudience,
  IntegrationAwsOidc,
  IntegrationStatusCode,
} from './types';

beforeEach(() => {
  jest.resetAllMocks();
});

test('fetch a single integration: fetchIntegration()', async () => {
  // test a valid response
  jest.spyOn(api, 'get').mockResolvedValue(awsOidcIntegration);

  let response =
    await integrationService.fetchIntegration<IntegrationAwsOidc>(
      'integration-name'
    );
  expect(api.get).toHaveBeenCalledWith(
    cfg.getIntegrationsUrl('integration-name')
  );
  expect(response).toEqual({
    kind: 'aws-oidc',
    name: 'aws-oidc-integration',
    resourceType: 'integration',
    details:
      'Enroll EC2, RDS and EKS resources or enable Web/CLI access to your AWS Account.',
    spec: {
      roleArn: 'arn-123',
      origin: undefined,
    },
    statusCode: IntegrationStatusCode.Running,
  });

  // test null response
  jest.spyOn(api, 'get').mockResolvedValue(null);

  response =
    await integrationService.fetchIntegration<IntegrationAwsOidc>(
      'integration-name'
    );
  expect(response).toEqual({
    resourceType: 'integration',
    statusCode: IntegrationStatusCode.Running,
    kind: undefined,
    name: undefined,
  });
});

test('fetch integration list: fetchIntegrations()', async () => {
  // test a valid response
  jest.spyOn(api, 'get').mockResolvedValue({
    items: [
      awsOidcIntegration,
      awsOidcIntegrationWithAudience,
      githubIntegration,
      nonAwsOidcIntegration,
    ],
    nextKey: 'some-key',
  });

  let response = await integrationService.fetchIntegrations();
  expect(api.get).toHaveBeenCalledWith(cfg.getIntegrationsUrl());
  expect(response).toEqual({
    nextKey: 'some-key',
    items: [
      {
        kind: 'aws-oidc',
        name: 'aws-oidc-integration',
        resourceType: 'integration',
        details:
          'Enroll EC2, RDS and EKS resources or enable Web/CLI access to your AWS Account.',
        spec: {
          roleArn: 'arn-123',
        },
        statusCode: IntegrationStatusCode.Running,
      },
      {
        kind: 'aws-oidc',
        name: 'aws-oidc-integration2',
        resourceType: 'integration',
        details:
          'Enroll EC2, RDS and EKS resources or enable Web/CLI access to your AWS Account.',
        spec: {
          roleArn: 'arn-12345',
          audience: 'aws-identity-center',
        },
        statusCode: IntegrationStatusCode.Running,
      },
      {
        kind: 'github',
        name: 'github-my-org',
        resourceType: 'integration',
        details: 'GitHub repository access for organization "my-org"',
        spec: { organization: 'my-org' },
        statusCode: IntegrationStatusCode.Running,
      },
      {
        kind: 'abc',
        name: 'non-aws-oidc-integration',
        resourceType: 'integration',
        statusCode: IntegrationStatusCode.Running,
      },
    ],
  });

  // test null response
  jest.spyOn(api, 'get').mockResolvedValue(null);

  response = await integrationService.fetchIntegrations();
  expect(response).toEqual({
    items: [],
    nextKey: undefined,
  });
});

test('fetchAwsDatabases response', async () => {
  // test a valid response
  jest
    .spyOn(api, 'post')
    .mockResolvedValue({ databases: mockAwsDbs, nextToken: 'next-token' });

  let response = await integrationService.fetchAwsRdsDatabases(
    'integration-name',
    'mysql',
    { region: 'us-east-1', nextToken: 'next-token' }
  );

  expect(response).toEqual({
    databases: [
      {
        engine: 'postgres',
        name: 'rds-1',
        uri: 'endpoint-1',
        status: 'Available',
        labels: [{ name: 'env', value: 'prod' }],
        accountId: 'account-id-1',
        resourceId: 'resource-id-1',
        vpcId: 'vpc-123',
        subnets: [],
        securityGroups: [],
      },
      {
        engine: 'mysql',
        name: 'rds-2',
        uri: 'endpoint-2',
        labels: [],
        status: undefined,
        accountId: undefined,
        resourceId: undefined,
        vpcId: undefined,
        subnets: [],
        securityGroups: [],
      },
      {
        engine: 'mysql',
        name: 'rds-3',
        uri: 'endpoint-3',
        labels: [],
        status: undefined,
        accountId: undefined,
        resourceId: undefined,
        vpcId: undefined,
        subnets: [],
        securityGroups: [],
      },
    ],
    nextToken: 'next-token',
  });

  // test null response
  jest.spyOn(api, 'post').mockResolvedValue(null);

  response = await integrationService.fetchAwsRdsDatabases(
    'integration-name',
    'mysql',
    {} as any
  );
  expect(response).toEqual({
    databases: [],
    nextToken: undefined,
  });
});

describe('fetchAwsDatabases() request body formatting', () => {
  test.each`
    protocol             | expectedEngines          | expectedRdsType
    ${'mysql'}           | ${['mysql', 'mariadb']}  | ${'instance'}
    ${'postgres'}        | ${['postgres']}          | ${'instance'}
    ${'aurora-mysql'}    | ${['aurora-mysql']}      | ${'cluster'}
    ${'aurora-postgres'} | ${['aurora-postgresql']} | ${'cluster'}
  `(
    'format protocol $protocol',
    async ({ protocol, expectedEngines, expectedRdsType }) => {
      jest.spyOn(api, 'post').mockResolvedValue({ databases: [] }); // not testing response here.

      await integrationService.fetchAwsRdsDatabases(protocol, protocol, {
        region: 'us-east-1',
        nextToken: 'next-token',
      });

      expect(api.post).toHaveBeenCalledWith(
        `/v1/webapi/sites/localhost/integrations/aws-oidc/${protocol}/databases`,
        {
          rdsType: expectedRdsType,
          engines: expectedEngines,
          region: 'us-east-1',
          nextToken: 'next-token',
        }
      );
    }
  );
});

test('fetch integration rules: fetchIntegrationRules()', async () => {
  // test a valid response
  jest.spyOn(api, 'get').mockResolvedValue({
    rules: [
      {
        resourceType: 'eks',
        region: 'us-west-2',
        labelMatcher: [{ name: 'env', value: 'dev' }],
        discoveryConfig: 'cfg',
        lastSync: 1733782634,
      },
    ],
    nextKey: 'some-key',
  });

  let response = await integrationService.fetchIntegrationRules(
    'name',
    AwsResource.eks
  );
  expect(api.get).toHaveBeenCalledWith(
    cfg.getIntegrationRulesUrl('name', AwsResource.eks)
  );
  expect(response).toEqual({
    nextKey: 'some-key',
    rules: [
      {
        resourceType: 'eks',
        region: 'us-west-2',
        labelMatcher: [{ name: 'env', value: 'dev' }],
        discoveryConfig: 'cfg',
        lastSync: 1733782634,
      },
    ],
  });

  // test null response
  jest.spyOn(api, 'get').mockResolvedValue(null);

  response = await integrationService.fetchIntegrationRules(
    'name',
    AwsResource.eks
  );
  expect(response).toEqual({
    nextKey: undefined,
    rules: [],
  });
});

const nonAwsOidcIntegration = {
  name: 'non-aws-oidc-integration',
  subKind: 'abc',
};

const awsOidcIntegration = {
  name: 'aws-oidc-integration',
  subKind: 'aws-oidc',
  awsoidc: { roleArn: 'arn-123' },
};

const awsOidcIntegrationWithAudience = {
  name: 'aws-oidc-integration2',
  subKind: 'aws-oidc',
  awsoidc: {
    roleArn: 'arn-12345',
    audience: IntegrationAudience.AwsIdentityCenter,
  },
};
const githubIntegration = {
  name: 'github-my-org',
  subKind: 'github',
  github: {
    organization: 'my-org',
  },
};

const mockAwsDbs = [
  {
    protocol: 'postgres',
    name: 'rds-1',
    uri: 'endpoint-1',
    labels: [{ name: 'env', value: 'prod' }],
    aws: {
      status: 'Available',
      account_id: 'account-id-1',
      rds: {
        resource_id: 'resource-id-1',
        vpc_id: 'vpc-123',
      },
    },
  },
  // Test with empty aws fields.
  {
    protocol: 'mysql',
    name: 'rds-2',
    uri: 'endpoint-2',
    aws: {},
  },
  // Test without aws field.
  {
    protocol: 'mysql',
    name: 'rds-3',
    uri: 'endpoint-3',
  },
];

File: lib/autoupdate/agent/integrations_test.go
/*
 * Teleport
 * Copyright (C) 2025  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package agent

import (
	"os"
	"path/filepath"
	"strings"
	"testing"

	"github.com/stretchr/testify/require"
)

func TestFindParentMatching(t *testing.T) {
	tests := []struct {
		name   string
		path   string
		parent string
		rpos   int
		result string
	}{
		{
			name:   "Has valid parent directory",
			path:   "/opt/teleport/default/versions/v1.2.3/bin/teleport",
			parent: "versions",
			rpos:   4,
			result: "/opt/teleport/default",
		},
		{
			name:   "Parent too close",
			path:   "/opt/teleport/default/versions/bin/teleport",
			parent: "versions",
			rpos:   4,
		},
		{
			name:   "Parent too far",
			path:   "/opt/teleport/default/versions/v1.2.3/extra/bin/teleport",
			parent: "versions",
			rpos:   4,
		},
		{
			name:   "Parent missing",
			path:   "/opt/teleport/default/version/v1.2.3/bin/teleport",
			parent: "versions",
			rpos:   4,
		},
		{
			name:   "Parent at root",
			path:   "/versions/v1.2.3/bin/teleport",
			parent: "versions",
			rpos:   4,
			result: "/",
		},
		{
			name:   "Position past root",
			path:   "/v1.2.3/bin/teleport",
			parent: "versions",
			rpos:   4,
		},
		{
			name: "Empty",
		},
		{
			name:   "Empty path",
			parent: "versions",
			rpos:   4,
		},
		{
			name:   "Empty parent",
			path:   "/v1.2.3/bin/teleport",
			parent: "",
			rpos:   4,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := findParentMatching(tt.path, tt.parent, tt.rpos)
			require.Equal(t, tt.result, result)
		})
	}
}

func TestStablePath(t *testing.T) {
	tests := []struct {
		name         string
		path         string
		resultIsFile bool
		resultIsLink bool
		result       string
		err          error
	}{
		{
			name:         "packaged path is file",
			path:         "/opt/teleport/system/bin/teleport",
			resultIsFile: true,
			result:       "/usr/local/bin/teleport",
		},
		{
			name:         "packaged path is link",
			path:         "/opt/teleport/system/bin/teleport",
			resultIsFile: true,
			resultIsLink: true,
			result:       "/usr/local/bin/teleport",
		},
		{
			name:         "packaged path is broken link",
			path:         "/opt/teleport/system/bin/teleport",
			resultIsLink: true,
			result:       "/opt/teleport/system/bin/teleport",
			err:          ErrUnstableExecutable,
		},
		{
			name:   "packaged path is missing",
			path:   "/opt/teleport/system/bin/teleport",
			result: "/opt/teleport/system/bin/teleport",
			err:    ErrUnstableExecutable,
		},
		{
			name:         "managed path is file",
			path:         "versions/version/bin/teleport",
			resultIsFile: true,
			result:       "[ns]/bin/teleport",
		},
		{
			name:         "managed path is link",
			path:         "versions/version/bin/teleport",
			resultIsFile: true,
			resultIsLink: true,
			result:       "[ns]/bin/teleport",
		},
		{
			name:         "managed path is broken link",
			path:         "versions/version/bin/teleport",
			resultIsLink: true,
			result:       "[ns]/versions/version/bin/teleport",
			err:          ErrUnstableExecutable,
		},
		{
			name:   "managed path is missing",
			path:   "versions/version/bin/teleport",
			result: "[ns]/versions/version/bin/teleport",
			err:    ErrUnstableExecutable,
		},
		{
			name:   "managed path is missing config",
			path:   "/test/versions/version/bin/teleport",
			result: "/test/versions/version/bin/teleport",
		},
		{
			name: "empty",
		},
		{
			name:   "other",
			path:   "/other",
			result: "/other",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			nsDir := t.TempDir()
			defaultPath := t.TempDir()
			createPath := defaultPath

			if tt.path != "" && !filepath.IsAbs(tt.path) {
				tt.path = filepath.Join(nsDir, tt.path)
				createPath = filepath.Join(nsDir, "bin")
				cfgPath := filepath.Join(nsDir, UpdateConfigName)
				err := writeConfig(cfgPath, &UpdateConfig{
					Version: UpdateConfigV1,
					Kind:    UpdateConfigKind,
					Spec: UpdateSpec{
						Path: createPath,
					},
				})
				require.NoError(t, err)
			}

			_, name := filepath.Split(tt.path)
			if tt.resultIsLink {
				filePath := filepath.Join(t.TempDir(), name)
				if tt.resultIsFile {
					createEmptyFile(t, filePath)
				}
				createSymlink(t, filePath, filepath.Join(createPath, name))
			} else if tt.resultIsFile {
				createEmptyFile(t, filepath.Join(createPath, name))
			}

			result, err := stablePathForBinary(tt.path, defaultPath)
			require.Equal(t, tt.result, strings.NewReplacer(
				defaultPath, "/usr/local/bin",
				nsDir, "[ns]",
			).Replace(result))
			require.Equal(t, tt.err, err)
		})
	}
}

func createEmptyFile(t *testing.T, path string) {
	t.Helper()
	err := os.MkdirAll(filepath.Dir(path), os.ModePerm)
	require.NoError(t, err)
	f, err := os.Create(path)
	require.NoError(t, err)
	err = f.Close()
	require.NoError(t, err)
}

func createSymlink(t *testing.T, oldname, newname string) {
	t.Helper()
	err := os.MkdirAll(filepath.Dir(newname), os.ModePerm)
	require.NoError(t, err)
	err = os.Symlink(oldname, newname)
	require.NoError(t, err)
}
